specs/marketplace.md

```diff
@@ -0,0 +1,381 @@
+
+If the proof is invalid, or slot was already filled by another node, then the transaction
+will revert, otherwise `SlotFilled(requestId, slotIndex)` event is emitted. If the transaction is successful, then the
+node SHOULD transition into __proving__ state as it will need to submit proof of data possession when prompted by the
+contract.
+
+It should be noted that if the node sees the `SlotFilled` emitted for slot that he is downloading the dataset or 
+generating proof for, then node SHOULD stop and choose different non-filled slot to try to fill as the chosen slot
+was filled by other node.
+
+### Proving
+
+Once a node fills a slot, it MUST periodically, yet non-deterministically provide proof to the smart contract that it 
+stores the data it should. Node SHOULD detect that proof is required using the `isProofRequired(slotId)` or that it will
+be required using the `willProofBeRequired(slotId)` in case the node is in [downtime](https://github.com/codex-storage/codex-research/blob/41c4b4409d2092d0a5475aca0f28995034e58d14/design/storage-proof-timing.md).
+
+Once node knows it has to provide a proof it MUST get the proof challenge using `getChallenge(slotId)` which then
+NEEDS to be incorporated into the proof generation as described in Proving RFC (**TODO: Proving RFC**).
+
+When proof is generated it MUST be submitted with a transaction calling `submitProof(slotId, proof)` function.
+
+#### Slashing
+
+There is a slashing scheme in place that is orchestrated by the smart contract to incentive correct behavior
+and proper proof submissions by the storage provider nodes. This scheme is configured on smart contract level and is 
+the same for all the participants in the network. The concrete values of this scheme can be obtained by `getConfig()` contract call.
+
+The slashing works in the following way:
+
+ - Node MAY miss at most `config.collateral.slashCriterion` proofs before it is slashed.
+ - It is then slashed `config.collateral.slashPercentage` percentage **of the originally asked collateral** (hence the slashing amount is always the same for the given request).
+ - If the number of times the node was slashed reaches above `config.collateral.maxNumberOfSlashes`, then the slot is freed, the remaining of node's collateral is burned and the slot is offered to other nodes for repair. Contract also emits the `SlotFreed(requestId, slotIndex)` event.
+
+If the number of concurrent freed slots reaches above the `request.ask.maxSlotLoss`, then the dataset is lost and the Request is failed. 
+The collateral of all the nodes that hosted Request's slots is burned and the event `RequestFailed(requestId)` is emitted.
+
+### Repair
+
+When slot is freed because of too many missed proofs, which SHOULD be detected by listening on the `SlotFreed(requestId, slotIndex)` event, then
```

Comment:
```suggestion
When a slot is freed because of too many missed proofs, which SHOULD be detected by listening on the `SlotFreed(requestId, slotIndex)` event, then
```